### PR TITLE
Add more privileges

### LINKED ISSUE #3173: add more privilege types
**Parent Issue**

- https://github.com/datafuselabs/databend/issues/2894

**Summary**

currently we have 5 privilege types:

- Usage
- Create
- Select
- Insert
- Set

we'd add more production-ready privilege types in the future. but right now, the up front high priority task is to setup the framework on privileges checking. we'd make a plan about which privileges to add first, and how to cover the privilege check logic into the current code path.

some thoughts:

- do not get the privilege types list too big
- make the semantics of the few privilege types clear (more comments?)
- cover the DML & DDL operations with the privilege checking logics first, add the query operations later

steps:

- [x] get the first-to-add privilege types list
- [x] check the behaviour on granting `SUPER` on a table object
- https://github.com/datafuselabs/databend/pull/3567

### PR DESCRIPTION
**Pull request by flaneur2020**
I hereby agree to the terms of the CLA available at: https://databend.rs/policies/cla/

## Summary

Summary about this PR

- adds the privileges which drafted in https://github.com/datafuselabs/databend/issues/3173#issuecomment-997934491
- parser about parsing the newly added privileges
- do not allow the admin privileges like SUPER, USAGE on a database or table
- tests on interpreter
- adapt the different ALL PRIVILEGES semantics between the different objects

## Changelog

- Improvement

## Related Issues

Fixes #3173

## Test Plan

Unit Tests

Stateless Tests

### CODE REVIEW

File: common/exception/src/exception.rs [hunk 1]
```diff
@@ -189,6 +189,7 @@ build_exceptions! {
     UnknownColumn(58),
     InvalidSourceFormat(59),
     StrParseError(60),
+    IllegalGrant(1144),
```
Comment (BohuTANG) [inline common/exception/src/exception.rs:192]: It would be nice order by Code(1144) number.

File: common/meta/types/src/user_grant.rs [hunk 2]
```diff
@@ -26,8 +27,26 @@ pub enum GrantObject {
     Table(String, String),
 }
 
+impl GrantObject {
+    pub fn allow_privilege(&self, privilege: UserPrivilegeType) -> bool {
```
Comment (BohuTANG) [inline common/meta/types/src/user_grant.rs:31]: Could you give some comments here to explain why we pass this check?

File: common/meta/types/src/user_privilege.rs [hunk 3]
```diff
@@ -25,35 +26,105 @@ use enumflags2::BitFlags;
...
             UserPrivilegeType::Set => "SET",
         })
     }
 }
 
+/// Each privilege has a set of contexts, which limits the grant objects which the privilege can
+/// take effects on, for example, we can not grant a SUPER privilege (which is Global level)
+/// to a table (in a table level).
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Hash)]
+pub enum UserPrivilegeContext {
```
Comment (BohuTANG) [inline common/meta/types/src/user_privilege.rs:95]: Snowflake's permission model has a clear hierarchy:
* Securable object: An entity to which access can be granted. Unless allowed by a grant, access will be denied.

* Role: An entity to which privileges can be granted. Roles are in turn assigned to users. Note that roles can also be assigned to other roles, creating a role hierarchy.

* Privilege: A defined level of access to an object. Multiple distinct privileges may be used to control the granularity of access granted.

* User: A user identity recognized by Snowflake, whether associated with a person or program.

https://docs.snowflake.com/en/user-guide/security-access-control-overview.html
 
Do we have the plan to split the UserPrivilegeContext to:
1. Object: Database, Table, Role, User
2. Roles: SystemAdmin

?

Comment (flaneur2020) [reply common/meta/types/src/user_privilege.rs:95]: Currently we have a GrantObject struct which resembles the Object in the snowflake model, and we have a plan to add the Role.

the UserPrivilegeContext mainly effects on the SHOW PRIVILEGES statement from mysql, it looks like:

https://dev.mysql.com/doc/refman/8.0/en/show-privileges.html

```
mysql> SHOW PRIVILEGES\G
*************************** 1. row ***************************
Privilege: Alter
  Context: Tables
  Comment: To alter the table
*************************** 2. row ***************************
Privilege: Alter routine
  Context: Functions,Procedures
  Comment: To alter or drop stored functions/procedures
*************************** 3. row ***************************
Privilege: Create
  Context: Databases,Tables,Indexes
  Comment: To create new databases and tables
*************************** 4. row ***************************
Privilege: Create routine
  Context: Databases
  Comment: To use CREATE FUNCTION/PROCEDURE
*************************** 5. row ***************************
Privilege: Create temporary tables
  Context: Databases
  Comment: To use CREATE TEMPORARY TABLE
...

```

besides this, the UserPrivilegeContext struct do not really have got many other benefits yet. I think we can skip the context field yet, until we found it nesscary in our privilege model.

here is our current model in planned:

- a GrantEntry has many privileges and works on an Object
- an Object can be Global, a Database or a Table (or Columns in the future)
- an User or Role has a GrantSet, which contains multiple grant entries
- an User can bind with multiple roles, the GrantEntries from the roles which binds to the user + the GrantEntries the user directly got, produces the effective grant set

Comment (BohuTANG) [reply common/meta/types/src/user_privilege.rs:95]: Got, thank you!

File: common/meta/types/src/user_grant.rs (hunk 2 above)
Comment (flaneur2020) [reply common/meta/types/src/user_grant.rs:31]: sure, thank you for reminding

**Comment by BohuTANG** [comment]
I think stateless tests `05_0006_ddl_grant_privilege` need some privileges:
```
--- /home/runner/work/databend/databend/tests/suites/0_stateless/05_0006_ddl_grant_privilege.result	2021-12-20 14:39:24.546288689 +0000
+++ /home/runner/work/databend/databend/tests/suites/0_stateless/05_0006_ddl_grant_privilege.stdout	2021-12-20 14:40:41.603672947 +0000

0: common_exception::exception::ErrorCode::IllegalGrant
             at common/exception/src/exception.rs:113:77
   1: common_meta_types::user_grant::GrantObject::validate_privileges
             at common/meta/types/src/user_grant.rs:42:24
   2: <databend_query::interpreters::interpreter_grant_privilege::GrantPrivilegeInterpreter as databend_query::interpreters::interpreter::Interpreter>::execute::{{closure}}::{{closure}}
             at query/src/interpreters/interpreter_grant_privilege.rs:53:9
```

File: common/exception/src/exception.rs (hunk 1 above)
Comment (flaneur2020) [reply common/exception/src/exception.rs:192]: do we try keep the error code compatibile with mysql or not?

if not, i think we can append it after `StrParseError(60)` as `IllegalGrant(61)`. therre's a "// uncategorized" category after the 600 code, but this error is a trivial sql error like `UnknownTable` I think

**Comment by flaneur2020** [comment]
```
mysql> grant ALL ON bendcloud TO 'test-grant';
Query OK, 0 rows affected (0.01 sec)

mysql> show grants for 'test-grant';
+---------------------------------------------------------------------+
| Grants for test-grant@%                                             |
+---------------------------------------------------------------------+
| GRANT USAGE ON *.* TO 'test-grant'@'%'                              |
| GRANT CREATE ON `bendcloud`.* TO 'test-grant'@'%'                   |
| GRANT ALL PRIVILEGES ON `bendcloud`.`bendcloud` TO 'test-grant'@'%' |
| GRANT CREATE ON `bendcloud`.`accounts` TO 'test-grant'@'%'          |
+---------------------------------------------------------------------+
4 rows in set (0.00 sec)

mysql> grant Super ON bendcloud TO 'test-grant';
ERROR 1144 (42000): Illegal GRANT/REVOKE command; please consult the manual to see which privileges can be used
mysql> grant create ON bendcloud TO 'test-grant';
Query OK, 0 rows affected (0.01 sec)

mysql> show grants for 'test-grant';
+---------------------------------------------------------------------+
| Grants for test-grant@%                                             |
+---------------------------------------------------------------------+
| GRANT USAGE ON *.* TO 'test-grant'@'%'                              |
| GRANT CREATE ON `bendcloud`.* TO 'test-grant'@'%'                   |
| GRANT ALL PRIVILEGES ON `bendcloud`.`bendcloud` TO 'test-grant'@'%' |
| GRANT CREATE ON `bendcloud`.`accounts` TO 'test-grant'@'%'          |
+---------------------------------------------------------------------+
4 rows in set (0.00 sec)

mysql> revoke create ON bendcloud FROM 'test-grant';
Query OK, 0 rows affected (0.01 sec)

mysql> show grants for 'test-grant';
+------------------------------------------------------------------------------------------------------------------------------------------------------+
| Grants for test-grant@%                                                                                                                              |
+------------------------------------------------------------------------------------------------------------------------------------------------------+
| GRANT USAGE ON *.* TO 'test-grant'@'%'                                                                                                               |
| GRANT CREATE ON `bendcloud`.* TO 'test-grant'@'%'                                                                                                    |
| GRANT SELECT, INSERT, UPDATE, DELETE, DROP, REFERENCES, INDEX, ALTER, CREATE VIEW, SHOW VIEW, TRIGGER ON `bendcloud`.`bendcloud` TO 'test-grant'@'%' |
| GRANT CREATE ON `bendcloud`.`accounts` TO 'test-grant'@'%'                                                                                           |
+------------------------------------------------------------------------------------------------------------------------------------------------------+
4 rows in set (0.00 sec)

mysql> GRANT all ON *.* TO 'test-grant';
Query OK, 0 rows affected (0.01 sec)

mysql> show grants for 'test-grant';
+------------------------------------------------------------------------------------------------------------------------------------------------------+
| Grants for test-grant@%                                                                                                                              |
+------------------------------------------------------------------------------------------------------------------------------------------------------+
| GRANT ALL PRIVILEGES ON *.* TO 'test-grant'@'%'                                                                                                      |
| GRANT CREATE ON `bendcloud`.* TO 'test-grant'@'%'                                                                                                    |
| GRANT SELECT, INSERT, UPDATE, DELETE, DROP, REFERENCES, INDEX, ALTER, CREATE VIEW, SHOW VIEW, TRIGGER ON `bendcloud`.`bendcloud` TO 'test-grant'@'%' |
| GRANT CREATE ON `bendcloud`.`accounts` TO 'test-grant'@'%'                                                                                           |
+------------------------------------------------------------------------------------------------------------------------------------------------------+
4 rows in set (0.00 sec)

mysql> revoke create ON *.* FROM 'test-grant';
Query OK, 0 rows affected (0.00 sec)

mysql> show grants for 'test-grant';
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Grants for test-grant@%                                                                                                                                                                                                                                                                                                                              |
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| GRANT SELECT, INSERT, UPDATE, DELETE, DROP, RELOAD, SHUTDOWN, PROCESS, FILE, REFERENCES, INDEX, ALTER, SHOW DATABASES, SUPER, CREATE TEMPORARY TABLES, LOCK TABLES, EXECUTE, REPLICATION SLAVE, REPLICATION CLIENT, CREATE VIEW, SHOW VIEW, CREATE ROUTINE, ALTER ROUTINE, CREATE USER, EVENT, TRIGGER, CREATE TABLESPACE ON *.* TO 'test-grant'@'%' |
| GRANT CREATE ON `bendcloud`.* TO 'test-grant'@'%'                                                                                                                                                                                                                                                                                                    |
| GRANT SELECT, INSERT, UPDATE, DELETE, DROP, REFERENCES, INDEX, ALTER, CREATE VIEW, SHOW VIEW, TRIGGER ON `bendcloud`.`bendcloud` TO 'test-grant'@'%'                                                                                                                                                                                                 |
| GRANT CREATE ON `bendcloud`.`accounts` TO 'test-grant'@'%'                                                                                                                                                                                                                                                                                           |
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
4 rows in set (0.00 sec)
```

it seems has different privilege set on GRANT ALL to diferent objects, needs more works on this semantic, convert back to draft 🤔

**Pull request merged**